### PR TITLE
Jetpack component & Modsuit module signal clean-up & fixes

### LINKED ISSUE #76116: Advanced ion jetpack module and normal ion jetpack module don't work
## Round ID:
[208399](https://scrubby.melonmesa.com/round/208399)

## Reproduction:

1. Print ion jetpack module for modsuit
2. Put it into modsuit
3. Go to space, turn ion jetpack on, realized its not working

Tested also advanced ion jetpack module, same result, you turn it on but its not working.

### PR DESCRIPTION
**Pull request by SyncIt21**
## About The Pull Request
1. Removed the `get_mover` callback, the mover can be retrieved during activation itself
2. Fixes #76116
the user is passed correctly during activation & deactivation same for modsuit modules and this also fixes the same bug for `/obj/item/organ/internal/cyberimp/chest/thrusters` as it's signal was also not registered correctly with the user
3. Timestop module on `on_module_triggered()` accepts user as 2nd param

## Changelog

:cl:
fix: jetpack modules work on mod suits again
fix: jetpack cyber implants also work
refactor: removed `get_mover` callback, user is retrieved during activation
refactor: timestop module on `on_module_triggered()` accepts user as 2nd param
/:cl:

### CODE REVIEW

**Comment by Sealed101** [comment]
i looked at how borgs and mechs do their thrusters/jetpacks and i weep

make sure it doesn't reintroduce whatever was happening with the jetpack signals when the jetpack gets unequipped and stored in a modsuit when you retract the modsuit chestplate. it's probably the modsuit retraction code that's at fault there really but still

**Comment by SyncIt21** [comment]
> when the jetpack gets unequipped and stored in a modsuit 

Screenshot (263)

Don't think that's a problem

Also when the jetpack get's unequiped it turns  itself off which in turn unregisters the signals
https://github.com/tgstation/tgstation/blob/7b7fa21e899b5d83b56acd949f7caf83b07e36d1/code/game/objects/items/tanks/jetpack.dm#L31-L34

**Comment by SyncIt21** [comment]
>  jetpack gets unequipped and stored in a modsuit when you retract the modsuit chestplate

How do i reproduce this?

**Comment by SyncIt21** [comment]
>  jetpack signals when the jetpack gets unequipped and stored in a modsuit when you retract the modsuit chestplate

That's fixed now because during deactivation the signals get unregistered with the correct user param passed

https://github.com/tgstation/tgstation/blob/c0a357629c8fa5accbb50bb7e116112b94b5c83b/code/modules/mod/modules/_module.dm#L139

**Comment by Stonetear** [comment]
I can't believe I broke jetpacks in #75844.  I had specifically asked someone to make sure my "fix" wasn't shitcode. :(

> make sure it doesn't reintroduce whatever was happening with the jetpack signals when the jetpack gets unequipped and stored in a modsuit when you retract the modsuit chestplate. it's probably the modsuit retraction code that's at fault there really but still

The issue was that by the time the unregister signal was called, the jetpack was already unequipped, so there was no user.

**Comment by JupiterJaeden** [comment]
Dear god speedmerge this please

File: code/__DEFINES/dcs/signals/signals_object.dm
Comment (LemonInTheDark) [inline code/__DEFINES/dcs/signals/signals_object.dm:306]: what are you trying to do here exactly? it's still used t he same why was it touched

**Comment by LemonInTheDark** [comment]
>     1. Removed the `get_mover` callback, the mover can be retrieved during activation itself

Good
 
>     2. Removed `stabilize` var from the component `Initialize` proc as all things that use this component(ion jetpack module, co2 jetpack, cyber implant jetpack) passed this var as FALSE anyway

Only really works with 3

>     3. Rather than adding the jetpack component every time stabilization is changed now it just adds the component once and changes the `stabilize` on the stored ref making toggling stabilization much faster & cleaner

Fundamentally disagree. Holding directly onto a ref to a component is HORRIBLE practice. This doesn't change speed significantly since jetpack dupemodes means the pack wasn't actually being created again, it was just ingesting new arguments. Don't do this.

>     4. Fixes [Advanced ion jetpack module and normal ion jetpack module don't work #76116](https://github.com/tgstation/tgstation/issues/76116)
>        the user is passed correctly during activation & deactivation same for modsuit modules and this also fixes the same bug for `/obj/item/organ/internal/cyberimp/chest/thrusters` as it's signal was also not registered correctly with the user

Based

>     5. Only 1 return value from the jetpack component when activation fails rather than passing a unique value for every case for consistency

Ok sure

>     6. Timestop module on `on_module_triggered()` accepts user as 2nd param

o7

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]

File: code/__DEFINES/dcs/signals/signals_object.dm
Comment (SyncIt21) [reply code/__DEFINES/dcs/signals/signals_object.dm:306]: Why have different returns values? If it fails let it just return one value that can be understood by everyone. 
Yes the previous one worked but it was just execess semantics sugar

**Comment by SyncIt21** [comment]
> Fundamentally disagree. Holding directly onto a ref to a component is HORRIBLE practice

Can you um elaborate more? I wasn't talking about weak refs or anything,  was just talking about a normal var that get's initialized once during AddComponent() and is used all over. How is that a bad thing?

File: code/__DEFINES/dcs/signals/signals_object.dm
Comment (SyncIt21) [reply code/__DEFINES/dcs/signals/signals_object.dm:306]: After all you just acknowledged this


> 5. Only 1 return value from the jetpack component when activation fails rather than passing a unique value for every case for consistency
> Ok sure

**Comment by LemonInTheDark** [comment]
> > Fundamentally disagree. Holding directly onto a ref to a component is HORRIBLE practice
> 
> Can you um elaborate more? I wasn't talking about weak refs or anything, was just talking about a normal var that get's initialized once during AddComponent() and is used all over. How is that a bad thing?

Components should manage their own state. if a component wants to stop existing, it should be allowed to do that.
Holding a ref to them denies them that purpose, and also just makes them end up a lot like just like helper datums.

They're meant to be in their own space, touching them externally is a code smell and we shouldn't be doing it like this.

File: code/__DEFINES/dcs/signals/signals_object.dm
Comment (LemonInTheDark) [reply code/__DEFINES/dcs/signals/signals_object.dm:306]: called isn't the right value here. though actually, I think I no longer agree with 2 hours ago me.
we want other consumers of these signals to be able to block them. They don't have them right now but we should support it, and I think that means we want a unique flag for each, if at least by name.

**Comment by SyncIt21** [comment]
That makes sense and all,  but for this specific use case it's just excess overhead, when we change a small boolean var called stabalize there is no need to Unregister/Reregister new signals,  reconsume all vars which is what we were doing in InheritComponent() of the jetpack component, when we can avoid it.

Also the jetpack component we know for this specific case does not stop existing by itself but only when its parent is deleted.

So while yes it is good to follow coding standards if we can optimize for speed in specific use cases like these,  that should take higher precedence.

**Comment by LemonInTheDark** [comment]
Speed isn't a concern here given how rarely this happens. Complexity maybe but speed is not worth considering. We are decidedly in userspace here.

**Comment by SyncIt21** [comment]
It's not rarely,  changing stabalize var is an UI option, when the jetpack get's turned off, you have to turn it back on again so its just as common as turning the jetpack on/off.

Also its not just for speed,  it improves code readability too, if we can store the component and document that var,  then people will know how to use it, which is also way better for debugging, since you know this var very well. Rather than digging through the components in the VV menu.

I don't know this feels like a very subjective discussion, your approach follows standards but has a performance cost, while mine is faster and also optimized for readability in this specific instance.

Sorry  but I'm just not sold on your narrative

File: code/__DEFINES/dcs/signals/signals_object.dm
Comment (SyncIt21) [reply code/__DEFINES/dcs/signals/signals_object.dm:309]: This define isn't tied to any particular signal, it's just a general failure for anything that tries to activate the jetpack component via any signal because if you look at the Initialize() proc you can see it is passed as a param `activation_signal`
https://github.com/tgstation/tgstation/blob/583538a12afcb5ef6e9a3b5e04d743b032569f3d/code/datums/components/jetpack.dm#L27

Which registers itself on the `activate` proc
https://github.com/tgstation/tgstation/blob/583538a12afcb5ef6e9a3b5e04d743b032569f3d/code/datums/components/jetpack.dm#L34

Which returns a failure if it occurs
https://github.com/tgstation/tgstation/blob/583538a12afcb5ef6e9a3b5e04d743b032569f3d/code/datums/components/jetpack.dm#L51-L52

So any future thing can pass in whatever signal via `activation_signal` and listen for the failure as follows
```
if(SEND_SIGNAL(src, <WHATEVER_SIGNAL>, user) & JETPACK_COMPONENT_ACTIVATION_FAILED)
   return //we have nothing to do here
```


So here the signal can be unique & whatever, & having a universal return value for failed activation creates a standard for eveyone to understand

**Comment by MrMelbert** [comment]
> It's not rarely, changing stabalize var is an UI option, when the jetpack get's turned off, you have to turn it back on again so its just as common as turning the jetpack on/off.

For the record when we're talking about speed and performance, it is really only applicable when it's called dozens of times a second. 

Stabilize setting is called once maybe every few minutes.

Speed is absolutely not a concern, you're vastly overestimating the cost of a single function call vs the grand scheme of things

**Comment by SyncIt21** [comment]
That's True, but I mean I already came this far with the code,  to put back all that jank is a bit too much work and we aren't talking about any visible consequences in game from this code(except for a very minute performance gain as you have pointed).

Even for readability I have added way fewer lines of code than I have removed 
Screenshot_20230626-153213_Chrome
so maintainence cost is far less.

Again I never disagreed with the points made, just saying this is a unique case where we can look past them without any future consequences

**Comment by LemonInTheDark** [comment]
I realize why you're against the idea, I just do not relish the pattern this creates or what it does to those signals. I think it measurably harms usability and future changes.
I don't think total work done should be a fair justification for a change, with all the fuckoff big prs I make I'd hate to think they're only merged because I'm holding the codebase hostage. 

Lines of code don't necessarily mean lower maintenance costs, and directly tying a component to its owner and removing space for non component users of a signal I think harms more then it helps.

I am concerned about this partially because like, I wrote it right, but also because I think it is worse in the long term, if only slightly.

**Comment by SyncIt21** [comment]
This creates a delima,  help me understand this.

For the jetpack component since we are changing a small boolean var inside this component it's easy to not hold onto a ref, that's fine 

But what about massive components like`/datum/components/remote_materials` OR `/datum/components/material_container`?
Whenever the material changes inside you don't see anywhere in code the component being readded infact it follows my approach of holding onto a ref of this component like this.
https://github.com/tgstation/tgstation/blob/f4557785bdb0bd0b0c16b805b1285cf6993319b3/code/modules/research/machinery/_production.dm#L38-L43

So why is it ok to hold onto ref there but not here? What's the defining boundary?

**Comment by SyncIt21** [comment]
@LemonInTheDark so umm, i don't know what do you want to do now? I will revert the changes no problem but only if i get a good explanation for the above comment.

**Comment by LemonInTheDark** [comment]
> This creates a delima, help me understand this.
>
> But what about components like`/datum/components/remote_materials` OR `/datum/components/material_container`? ...
>
> So why is it ok to hold onto a ref there but not for my jetpack component here? What's the defining boundary?

Those components suck and are bad. They're only sort of componentizing behavior (moving out reacting to multitools and such) and instead serve on the whole as datums that "owners" can act on.

This is a bad match for the component system (The core idea of which is separate bits of data and behavior that react to "events", signals, happening to their parent and the world around them), and leads to a lot of stinky reference holding and direct invocation.
They should instead be implemented as just bare /datum's, like mood, alarms, storage, that sort of thing. They don't want to be components, and they chaff on the boundaries of the system.

It isn't ok to do what these are doing, I would have the same qualms about them that I do about this. They're way worse, in a way that makes their core concept unfit for componentization that does not apply here. Shouldn't look to them as good examples tho, cause again they kinda just suck.

**Comment by SyncIt21** [comment]
Ahh, i see so we made stuff like material containers & remote materials into components by mistake. That's fair. I did learn to code by looking at examples like this from the code base thinking what they did was correct not realizing they themselves messed up like the rest of us.

I guess i could make a future PR that turns material container into a regular datums and not `datum/component` for some free GBP.

Thanks for debating with me for this long, I could have just made the changes you requested a week back but i always want to understand the methodology behind such changes first rather than blindly implement them, like for example you gave a good reason why using `src` inside `ui_act` is bad in https://github.com/tgstation/tgstation/pull/74696#discussion_r1166099547 but then you quickly went back on your statement https://github.com/tgstation/tgstation/pull/76007#issuecomment-1588164150 -_- so i kinda lost confidence in your reasoning. 

> I think

Also when you use terms like these i feel like you yourself aren't sure what you are talking about so you know, i too started questioning my life choices but now i have no doubts

I'll revert the changes

**Comment by LemonInTheDark** [comment]
> Thanks for debating with me for this long, I could have just made the changes you requested a week back but i always want to understand the methodology behind such changes first rather than blindly implement them, like for example you gave a good reason why using `src` inside `ui_act` is bad in [#74696 (comment)](https://github.com/tgstation/tgstation/pull/74696#discussion_r1166099547) but then you quickly went back on your statement [#76007 (comment)](https://github.com/tgstation/tgstation/pull/76007#issuecomment-1588164150) -_- so i kinda lost confidence in your reasoning.

TBH I just like, forgot that callbacks can carry usr over? Like the concern makes sense if they can't, but they can so it's sorta pointless to worry about. Shoulda been more definite on that, sorry for the confusion.

> Also when you use terms like this i feel like you yourself aren't sure what you are talking about so you know, i too started questioning my life choices but now i have no doubts

I am always open to being stupendously wrong, though I realize that doesn't always generate an amazing amount of confidence.
Was using the term incorrectly here, u right.

On the I think topic, and to maybe introduce some doubts (and nuance).
The pattern of holding direct refs to components is one that not everyone is as against as I am. Pretty much everyone hates "datum like" components, but you will find maintainer code (singularity is a good example) that involves holding `/datum/weakref`s to components and such (I think weakrefs are best if you're gonna do it, because of all that stuff about components managing and deleting themselves)

Most of those examples are "it can't be done nicely any other way" because a thing is treading the line of what a component is for like, gameplay reasons and such.
But like this pr got brought up on discord and moth and melbert both talked about how they do this sort of thing sometimes. I am very much on the absolutist side of this debate.

If you can do without it that is best, but most of the other maints aren't nearly as hardline on this topic as me, and I don't want to mislead you on that.

Like it **is** best to avoid, but I **think** it's always avoidable if that makes any sense.

**Comment by SyncIt21** [comment]
Can't be mad, the game is the game.

This has been pending long enough, could you merge this asap if possible?

**Comment by SyncIt21** [comment]
1 last question regarding all of this. Since you are against holding ref to components why 

1. Does `AddComponent() return a ref to the component added? Doesn't that just encourage people to make a ref to hold the return value
2. Why is there a proc called `GetComponent()` which again returns a ref to a component?

In the future do you plan to change these or are there exceptions where they are still useful?

**Comment by LemonInTheDark** [comment]
>     1. Does `AddComponent()` return a ref to the component added? Doesn't that just encourage people to make a ref to hold the return value?

Yes
 
>     2. Why is there a proc called `GetComponent()` which again returns a ref to a component?

Because we had and have shit like the mat storage component that you need to reach out and touch, and components were born from a mix of goals, and developed over time.
S cluttered, and opinionated but not as strongly as I am.

> In the future do you plan to change these or are there exceptions where they are still useful?

In a perfect world I would, but there are cases where you don't want to make a full on component (singularity comp comes to mind), and maintainers who disagree with me on the hardline part of my stance.
Maybe if I can convince them or somethin idk man.

File: code/__DEFINES/dcs/signals/signals_object.dm
Comment (LemonInTheDark) [inline code/__DEFINES/dcs/signals/signals_object.dm:309]: I'm sorry I know this is taking forever, but I still really don't like the smell of this.
Signals are supposed to be like, generic right? in theory anyone could receive these and say "hey no you can't turn your (jetpack/thruster/modsuit pack) on right now"

I don't like this change because it removes that for cleanliness now, but ends up tightly tying the component to these 3 signals, which impedes future users. (and also leaves the other flags just sitting there which I think was a mistake).

Comment (SyncIt21) [reply code/__DEFINES/dcs/signals/signals_object.dm]: It's done. Please lets end this

**Comment by SyncIt21** [comment]
Unrelated question to this PR, is there a way to organize all includes inside`tgstation.dme` file in VS Code so it doesn't fail linters?

**Comment by LemonInTheDark** [comment]
finally free

**Pull request merged**
